### PR TITLE
[ONNX] Support logsoftmax with dim != -1

### LINKED ISSUE #17918: Allow ONNX export of Softmax with dim != -1 (including Softmax2d)
## 🚀 Feature

Right now, it is not possible to export a softmax function that doesn't use `dim=-1`. It should be possible to use softmax with arbitrary dimensions without the use of hacky workarounds by the user. 

## Motivation

There are lots of applications that require pixel-level predictions, such as segmentation and detection. 
I personally use the ONNX export function all the time, because ONNX is the most flexible when moving between frameworks and for deploying. All my models are typically fully convolutional and most of them use the softmax activation.

## Pitch

Allow transparent export of softmax layers with arbitrary `dim` parameter to ONNX. Use tensor transposition and reshaping only when necessary, without requiring the user to do it themselves.

## Alternatives

https://github.com/pytorch/pytorch/blob/master/torch/onnx/symbolic.py#L577 mentions that the reason for this limitation is that the semantics of PyTorch's `dim` and ONNXs `axis` are different. There are workarounds, such as transposing the channel dimension with the softmax dimension and transposing back after that. But that feels very hacky, especially when there are such functions as `Softmax2d`. 

Something like that might work:
```
class ONNXCompatibleSoftmax(nn.Softmax):

    def __init__(self, dim):
        super(ONNXCompatibleSoftmax, self).__init__(dim=-1)

        self._dim = dim
        self.softmax = super(ONNXCompatibleSoftmax, self).forward

    def forward(self, x):
        return torch.transpose(self.softmax(torch.transpose(x, 1, self._dim)), 1, self._dim)
```

cc @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

### PR DESCRIPTION
**Pull request by BowenBao**
PyTorch dim and ONNX axis have different meanings.
ONNX only supports log_softmax with dim = -1. Transpose must be added before and after log_softmax to support other cases.
This requires input rank to be known at export time.
Fixes https://github.com/pytorch/pytorch/issues/17918

### CODE REVIEW

**Review by lara-hdr — APPROVED** [review]
LGTM

**Review by houseroad — APPROVED** [review]
Looks good.
